tests/src/test/java/org/eclipse/hono/tests/amqp/AdapterAmqpIT.java

```diff
@@ -0,0 +1,407 @@
+                                        .closeHandler(context.asyncAssertFailure(expected -> {
+                                            // THEN the adapter rejects the request and closes the link.
+                                            completionTracker.complete();
+                                        })).open();
+                            }
+                        }).open();
+            } else {
+                context.fail(ar.cause());
+            }
+        });
+        completionTracker.awaitSuccess();
+    }
+
+    /**
+     * Verifies that a request to upload a message to an invalid endpoint results in the adapter closing the link used
+     * for sending the message.
+     * 
+     * @param context The context for running asynchronous test on Vert.x
+     */
+    @Test
+    public void testUploadMessageToInvalidEndpoint(final TestContext context) {
+        final Async completionTracker = context.async();
+
+        final String tenantId = helper.getRandomTenantId();
+        final String deviceId = helper.getRandomDeviceId(tenantId);
+        final ResourceIdentifier resource = ResourceIdentifier.from("UNKNOWN", tenantId, deviceId);
+        client = ProtonClient.create(vertx);
+        client.connect(IntegrationTestSupport.AMQP_HOST, IntegrationTestSupport.AMQP_PORT, ar -> {
+            if (ar.succeeded()) {
+                ar.result().openHandler(remoteOpen -> {
+                    if (remoteOpen.succeeded()) {
+                        final ProtonConnection openedConnection = remoteOpen.result();
+                        openedConnection.createSender(getTargetAddress(resource))
+                                .closeHandler(context.asyncAssertFailure(failed -> {
+                                    completionTracker.complete();
+                                })).open();
+                    }
+                }).open();
+            } else {
+                context.fail(ar.cause());
+            }
+        });
+        completionTracker.awaitSuccess();
+    }
+
+    // ------------------------------< private methods >---
+    private void doUploadMessages(final TestContext context, final ResourceIdentifier resource)
+            throws InterruptedException {
+        final CountDownLatch received = new CountDownLatch(MESSAGES_TO_SEND);
+        final Async consumerTracker = context.async();
+        createConsumer(resource.getTenantId(), Boolean.TRUE, msg -> {
+            LOG.info("received message {}", msg);
+            received.countDown();
+
+        }).setHandler(context.asyncAssertSuccess(ok -> consumerTracker.complete()));
+        consumerTracker.await();
+
+        final AtomicInteger messageCount = new AtomicInteger(0);
+        client = ProtonClient.create(vertx);
+        while (messageCount.get() < MESSAGES_TO_SEND) {
+            final Async sending = context.async();
+            client.connect(IntegrationTestSupport.AMQP_HOST, IntegrationTestSupport.AMQP_PORT, ar -> {
```

Comment:
why do you want to connect for each message individually?

Reply:
otherwise the consumer receives a single message.. trying to understand why that is so